bindings/engine.i

```diff
@@ -129,19 +129,20 @@ static const GncGUID * gncBudgetGetGUID(GncBudget *x)
 SplitsVec gnc_get_match_commodity_splits (AccountVec accounts, bool use_end_date,
                                           time64 end_date, gnc_commodity *comm, bool sort)
 {
-    auto match = [use_end_date, end_date, comm](const Split* s) -> bool
+    SplitsVec rv;
+    auto maybe_accumulate = [&rv, use_end_date, end_date, comm](auto s) -> bool
     {
         if (xaccSplitGetReconcile (s) == VREC) return false;
         auto trans{xaccSplitGetParent (s)};
-        if (use_end_date && xaccTransGetDate(trans) > end_date) return false;
+        if (use_end_date && xaccTransGetDate(trans) > end_date) return true;
         auto txn_comm{xaccTransGetCurrency (trans)};
         auto acc_comm{xaccAccountGetCommodity (xaccSplitGetAccount (s))};
-        return (txn_comm != acc_comm) && (!comm || comm == txn_comm || comm == acc_comm);
+        if ((txn_comm != acc_comm) && (!comm || comm == txn_comm || comm == acc_comm))
+            rv.push_back (const_cast<Split*>(s));
+        return false;
     };
-    std::vector<Split*> rv;
-    auto maybe_accumulate_split = [&rv, match](auto s){ if (match(s)) rv.push_back (s); };
-    for (const auto acc : accounts)
-        gnc_account_foreach_split (acc, maybe_accumulate_split, true);
+    auto scan_account = [&](auto acc){ gnc_account_find_split (acc, maybe_accumulate, false); };
```

Comment:
Yuk. Passing a function with side effects to `gnc_account_find_split`'s predicate is ugly and obscure. 

To correctly indicate what you want we need a `gnc_account_foreach_split_until` function. Unfortunately I think you need C++20's ranges library to implement that efficiently with standard algorithms. A proper use of std::find_if and std::accumulate or std::for_each would loop over the desired part of the vector twice, once to find the last desired split and the second to do the accumulation, so you'd have to write a loop with an escape, something like 
```
template <typename IterType, typename FuncType, typename PredType> void
foreach_until(IterType begin, IterType end, FuncType func, PredType pred)
/* Insert appropriate enable_ifs to constrain the types */
{
    for (auto local_iter = begin; local_iter != end && pred(local_iter); ++local_iter)
       func(local_iter);
}
```

Reply:
In this case it'd be `std::upper_bound` to find the date, then `std::for_each`. Further Account.h methods would be needed.

Reply:
Except that std::upper_bound and friends work with forward iterators only and you want to be able to use reverse iterators.

Reply:
IMO this limitation is acceptable...